rs/xnet/payload_builder/src/lib.rs

```diff
@@ -903,6 +925,16 @@ pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usi
     }
 }
 
+/// The stream index up to which messages can be inducted while limiting the
+/// number of signals in the reverse stream to `MAX_STREAM_MESSAGES`.
+///
+/// `stream_begin` is the `begin` in the `StreamHeader` contained in the (same) stream slice.
+///  It reflects the status on the remote subnet as far as we know at present. Up to this index
+///  signals can be gc'ed in the reverse stream.
+pub fn max_message_index(stream_begin: StreamIndex) -> StreamIndex {
+    stream_begin + (MAX_STREAM_MESSAGES as u64).into()
+}
```

Comment:
I was just wondering what would happen if we got a malicious `stream_begin` here, i.e. what if this wraps around?
First of all it would mean that `max_message_index` << `messages_begin` in which case we just wouldn't include any messages in slices.
But there are also sanity checks for this in `validate_slice()`, for [begin and end in the header](https://sourcegraph.com/github.com/dfinity/ic@5c85e7d4570d9636f896ec6fdb0e507ff4e01d08/-/blob/rs/xnet/payload_builder/src/lib.rs?L619) and also to ensure `messages_begin` is [within this interval](https://sourcegraph.com/github.com/dfinity/ic@5c85e7d4570d9636f896ec6fdb0e507ff4e01d08/-/blob/rs/xnet/payload_builder/src/lib.rs?L619).

So unless the stream indices are very close to `u64::MAX` in which case all kinds of other stuff would also break, this is not an issue. We'll never get there though; it would take thousands of years.

Reply:
This is one of the reasons why I was arguing for using `ExpectedInfices` for this, i.e. the value that we know to be correct, as opposed to a value fed to us by a potentially malicious subnet.

Even though the latter is probably validated again and again, including in the block maker, so by the time it gets to us it's just as safe to use. But (to some extent) it removes the need to wonder about "what if this index was invalid".